test/e2e_node/restart_test.go

```diff
@@ -265,5 +270,117 @@ var _ = SIGDescribe("Restart [Serial] [Slow] [Disruptive]", func() {
 				}
 			}
 		})
+		// Regression test for https://issues.k8s.io/116925
+		ginkgo.It("should delete pods which are marked as terminal and have a deletion timestamp set after restart", func(ctx context.Context) {
```

Comment:
e2e serial run said this took only 5 seconds to complete.  That's... suspiciously fast.  Too fast?

Reply:
I updated the e2e test to remove the `sleep 5` from the SIGTERM handler so the container will exit right away (with success exit code) on SIGTERM. Here's the build logs, everything looks WAI:

```
[sig-node] Restart [Serial] [Slow] [Disruptive] Kubelet should delete pods which are marked as terminal and have a deletion timestamp set after restart
test/e2e_node/restart_test.go:274
  STEP: Creating a kubernetes client @ 04/11/23 09:42:21.496
  STEP: Building a namespace api object, basename restart-test @ 04/11/23 09:42:21.496
  Apr 11 09:42:21.500: INFO: Skipping waiting for service account
  STEP: Creating a pod (restart-test-806/terminal-restart-pode7358abd-c6a2-4823-87b9-5dc93abe9650) with restart policy: Never @ 04/11/23 09:42:21.5
  STEP: Waiting for the pod (restart-test-806/terminal-restart-pode7358abd-c6a2-4823-87b9-5dc93abe9650) to be running @ 04/11/23 09:42:21.504
  STEP: Deleting the pod (restart-test-806/terminal-restart-pode7358abd-c6a2-4823-87b9-5dc93abe9650) to set a deletion timestamp @ 04/11/23 09:42:23.512
  STEP: Started watch for pod (restart-test-806/terminal-restart-pode7358abd-c6a2-4823-87b9-5dc93abe9650) to enter succeeded phase @ 04/11/23 09:42:24.525
  I0411 09:42:24.525620    1061 retrywatcher.go:247] Starting RetryWatcher.
  STEP: Observed Pod (restart-test-806/terminal-restart-pode7358abd-c6a2-4823-87b9-5dc93abe9650) in phase Running @ 04/11/23 09:42:24.528
  STEP: Found Pod (restart-test-806/terminal-restart-pode7358abd-c6a2-4823-87b9-5dc93abe9650) in phase Succeeded @ 04/11/23 09:42:24.747
  STEP: Ended watch for pod entering succeeded phase @ 04/11/23 09:42:24.747
  STEP: Stopping the kubelet @ 04/11/23 09:42:24.747
  I0411 09:42:24.747491    1061 retrywatcher.go:147] "Stopping RetryWatcher."
  I0411 09:42:24.747584    1061 retrywatcher.go:281] Stopping RetryWatcher.
  Apr 11 09:42:24.761: INFO: Get running kubelet with systemctl:   UNIT                            LOAD   ACTIVE SUB     DESCRIPTION
    kubelet-20230411T085234.service loaded active running /tmp/node-e2e-20230411T085234/kubelet --kubeconfig /tmp/node-e2e-20230411T085234/kubeconfig --root-dir /var/lib/kubelet --v 4 --hostname-override n1-standard-2-cos-stable-105-17412-1-61-0b523e8d --container-runtime-endpoint unix:///run/containerd/containerd.sock --config /tmp/node-e2e-20230411T085234/kubelet-config --feature-gates=DisableKubeletCloudCredentialProviders=true,KubeletCredentialProviders=true --image-credential-provider-config=/tmp/node-e2e-20230411T085234/credential-provider.yaml --image-credential-provider-bin-dir=/tmp/node-e2e-20230411T085234 --kernel-memcg-notification=true --cgroup-driver=systemd --cgroups-per-qos=true --cgroup-root=/ --runtime-cgroups=/system.slice/containerd.service

  LOAD   = Reflects whether the unit definition was properly loaded.
  ACTIVE = The high-level unit activation state, i.e. generalization of SUB.
  SUB    = The low-level unit activation state, values depend on unit type.
  1 loaded units listed.
  , kubelet-20230411T085234
  W0411 09:42:24.792008    1061 util.go:477] Health check on "http://127.0.0.1:10248/healthz" failed, error=Head "http://127.0.0.1:10248/healthz": dial tcp 127.0.0.1:10248: connect: connection refused
  STEP: Starting the kubelet @ 04/11/23 09:42:24.792
  W0411 09:42:24.809825    1061 util.go:477] Health check on "http://127.0.0.1:10248/healthz" failed, error=Head "http://127.0.0.1:10248/healthz": dial tcp 127.0.0.1:10248: connect: connection refused
  STEP: After the kubelet is restarted, verify the pod (restart-test-806/terminal-restart-pode7358abd-c6a2-4823-87b9-5dc93abe9650) is deleted by kubelet @ 04/11/23 09:42:26.813
  Apr 11 09:42:26.815: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
  STEP: Destroying namespace "restart-test-806" for this suite. @ 04/11/23 09:42:26.817
• [5.323 seconds]
```

And from the kubelet side, the pod is being handled correctly after restart: (Kubelet logs- https://gist.githubusercontent.com/bobbypage/f833a8a9bbfcb4e6af15b0953e560e89/raw/e7a82c0fca916dbbbaee0f799ccda7c131598127/gistfile0.txt)

Pod added for first time:

```
I0411 09:42:21.503828  141575 kubelet.go:2343] "SyncLoop ADD" source="api" pods=[restart-test-806/terminal-restart-pode7358abd-c6a2-4823-87b9-5dc93abe9650]
```

Phase is updated to succeeded:

Delete is seen
```
I0411 09:42:24.515610  141575 kubelet.go:2359] "SyncLoop DELETE" source="api" pods=[restart-test-806/terminal-restart-pode7358abd-c6a2-4823-87b9-5dc93abe9650]
```

SyncTerminatingPod runs:

```
I0411 09:42:24.515776  141575 kubelet.go:1943] "SyncTerminatingPod enter" pod="restart-test-806/terminal-restart-pode7358abd-c6a2-4823-87b9-5dc93abe9650" podUID=b9b11ae3-4bcc-4ccc-937a-fa355e1c1866
```

```
I0411 09:42:24.734241  141575 kubelet.go:2029] "SyncTerminatingPod exit" pod="restart-test-806/terminal-restart-pode7358abd-c6a2-4823-87b9-5dc93abe9650" podUID=b9b11ae3-4bcc-4ccc-937a-fa355e1c1866
```

Status is updated to succeeded 
```
I0411 09:42:24.746815  141575 status_manager.go:839] "Status for pod updated successfully" pod="restart-test-806/terminal-restart-pode7358abd-c6a2-4823-87b9-5dc93abe9650" statusVersion=4 status={Phase:Succeeded Conditions:[{Type:Initialized Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2023-04-11 09:42:21 +0000 UTC Reason:PodCompleted Message:} {Type:Ready Status:False LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2023-04-11 09:42:24 +0000 UTC Reason:PodCompleted Message:} {Type:ContainersReady Status:False LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2023-04-11 09:42:24 +0000 UTC Reason:PodCompleted Message:} {Type:PodScheduled Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2023-04-11 09:42:21 +0000 UTC Reason: Message:}] Message: Reason: NominatedNodeName: HostIP:10.138.0.42 PodIP: PodIPs:[] StartTime:2023-04-11 09:42:21 +0000 UTC InitContainerStatuses:[] ContainerStatuses:[{Name:terminal-restart-pode7358abd-c6a2-4823-87b9-5dc93abe9650 State:{Waiting:nil Running:nil Terminated:&ContainerStateTerminated{ExitCode:0,Signal:0,Reason:Completed,Message:,StartedAt:2023-04-11 09:42:21 +0000 UTC,FinishedAt:2023-04-11 09:42:24 +0000 UTC,ContainerID:containerd://909f4ece6fbf3d2920dc984354f79a94e598e77d583ba334d7cf24fa2dffb231,}} LastTerminationState:{Waiting:nil Running:nil Terminated:nil} Ready:false RestartCount:0 Image:registry.k8s.io/e2e-test-images/busybox:1.29-4 ImageID:registry.k8s.io/e2e-test-images/busybox@sha256:2e0f836850e09b8b7cc937681d6194537a09fbd5f6b9e08f4d646a85128e8937 ContainerID:containerd://909f4ece6fbf3d2920dc984354f79a94e598e77d583ba334d7cf24fa2dffb231 Started:0xc0011cb439 AllocatedResources:map[] Resources:nil}] QOSClass:BestEffort EphemeralContainerStatuses:[] Resize:}
```

Kubelet restarts due to e2e test triggering restart as due to pod entering terminal phase:

```
I0411 09:42:24.874499  141807 flags.go:64] FLAG: --address="0.0.0.0"
```


Pod is added again after restart:

```
I0411 09:42:25.894460  141807 kubelet.go:2343] "SyncLoop ADD" source="api" pods=[restart-test-806/terminal-restart-pode7358abd-c6a2-4823-87b9-5dc93abe9650]
I0411 09:42:25.894582  141807 pod_workers.go:770] "Pod is being synced for the first time" pod="restart-test-806/terminal-restart-pode7358abd-c6a2-4823-87b9-5dc93abe9650" podUID=b9b11ae3-4bcc-4ccc-937a-fa355e1c1866 updateType="create"
```

Pod is deleted finally:

```
I0411 09:42:25.908268  141807 status_manager.go:868] "Pod fully terminated and removed from etcd" pod="restart-test-806/terminal-restart-pode7358abd-c6a2-4823-87b9-5dc93abe9650"
```

Reply:
Agree, it looks fast, but it looks WAI, I guess it's good see latency is down (even considering a kubelet restart in the middle of this test)!

Reply:
Nice! It seems we should create a follow up Issue / PR to apply the same strategy to other tests which use sleep wait for SIGTERM.  Searching by "do sleep 5" I get hits in deleted_pods_test.go, node_shutdown_test.go, eviction_test.go, and system_node_critical_test.go

Reply:
Created followup issue for that - https://github.com/kubernetes/kubernetes/issues/117367